vertical-pod-autoscaler/pkg/admission-controller/logic/server_test.go

```diff
@@ -0,0 +1,284 @@
+						"containers": [{}]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("1"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourcesPatch(0),
+				addRequestsPatch(0),
+				addResourceRequestPatch(0, cpu, "1"),
+				addAnnotationRequest([][]string{{cpu}}),
+			},
+			nil,
+		},
+		{
+			"replacement cpu recommendation",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"resources": {
+									"requests": {
+										"cpu": "0"
+									}
+								}
+							}
+						]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("1"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourceRequestPatch(0, cpu, "1"),
+				addAnnotationRequest([][]string{{cpu}}),
+			},
+			nil,
+		},
+		{
+			"two replacement resources",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"resources": {
+									"requests": {
+										"cpu": "0"
+									}
+								}
+							}
+						]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu:        resource.MustParse("1"),
+						unobtanium: resource.MustParse("2"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourceRequestPatch(0, cpu, "1"),
+				addResourceRequestPatch(0, unobtanium, "2"),
+				addAnnotationRequest([][]string{{cpu, unobtanium}}),
+			},
+			nil,
+		},
+		{
+			"two containers",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"resources": {
+									"requests": {
+										"cpu": "0"
+									}
+								}
+							},
+							{}
+						]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("1"),
+					},
+				},
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("2"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourceRequestPatch(0, cpu, "1"),
+				addResourcesPatch(1),
+				addRequestsPatch(1),
+				addResourceRequestPatch(1, cpu, "2"),
+				addAnnotationRequest([][]string{{cpu}, {cpu}}),
+			},
+			nil,
+		},
+	}
+	for _, tc := range tests {
+		fppp := fakePodPreProcessor{e: tc.preProcessorError}
+		frp := fakeRecommendationProvider{tc.recommendResources, tc.recommendAnnotations, tc.recommendName, tc.recommendError}
+		s := NewAdmissionServer(&frp, &fppp)
+		patches, err := s.getPatchesForPodResourceRequest(tc.podJson, tc.namespace)
+		if (err == nil) != (tc.expectError == nil) || (err != nil && err.Error() != tc.expectError.Error()) {
+			t.Errorf("[%s] Expected error %v, got %v", tc.name, tc.expectError, err)
```

Comment:
I suggest  using https://godoc.org/github.com/stretchr/testify/assert for assertions.

Reply:
Done.